Update MicroProfile REST Client to 4.0

avpinchuk opened the pull request.

avpinchuk committed:
Update MicroProfile REST Client to 4.0

Signed-off-by: Alexander Pinčuk <alexander.v.pinchuk@gmail.com>

avpinchuk committed:
Update MicroProfile REST Client TCK to 4.0

Signed-off-by: Alexander Pinčuk <alexander.v.pinchuk@gmail.com>

OndroMih reviewed (commented):
Does it pass the MP Rest Client 4.0 TCK?

avpinchuk commented:
> Does it pass the MP Rest Client 4.0 TCK?

All but one test. The `ClientClosedTest` gives the deployment error because test war contains three ambiguous endpoins.

avpinchuk commented on appserver/distributions/glassfish/src/main/patches/microprofile-rest-client-api/META-INF/MANIFEST.MF:
```diff
@@ -1,42 +1,44 @@
- lient.git/microprofile-rest-client-api",tag="3.0"
-Build-Jdk-Spec: 1.8
-Created-By: 1.8.0_292 (AdoptOpenJDK)
+Bundle-SCM: url="https://github.com/eclipse/microprofile-rest-client/mic
```
Sure

OndroMih commented:
Well, does it make sense to raise a challenge?

OpenLiberty seems to be OK with it. I haven't run it against the TCK but here's their runner: https://github.com/OpenLiberty/open-liberty/blob/integration/dev/io.openliberty.microprofile.rest.client.4.0.internal_fat_tck/publish/tckRunner/tck/pom.xml

WildFly seems to implement it and pass the TCK too. Here's their runner: https://github.com/wildfly/wildfly/blob/main/testsuite/integration/microprofile-tck/rest-client/pom.xml

dmatej commented:
> > Does it pass the MP Rest Client 4.0 TCK?
> 
> All but one test. The `ClientClosedTest` gives the deployment error because test war contains three ambiguous endpoins.

If the test is using Arquillian, we already know where is the problem probably - in the OmniFish's gf container we leak rest clients, @arjantijms reproduced some failure on the CI. It leads to OOME sometimes and few other weird issues. It is possible that it is not the last thing to fix ...

avpinchuk commented:
> If the test is using Arquillian, we already know where is the problem probably - in the OmniFish's gf container we leak rest clients, @arjantijms reproduced some failure on the CI. It leads to OOME sometimes and few other weird issues. It is possible that it is not the last thing to fix ...

You're right. It works without OOME only with `-Xmx4G`.

avpinchuk commented:
> Well, does it make sense to raise a challenge?

I don't know. But if we try to deploy `_DEFAULT___DEFAULT__ClientClosedTest.war` on fresh GF server manually, we get an error:

```
[2025-08-27T00:25:33.216023+03:00] [GF 7.1.0-SNAPSHOT] [SEVERE] [] [jakarta.enterprise.system.core] [tid: _ThreadID=73 _ThreadName=admin-listener(1)] [level
Value: 1000] [[
  Exception while loading the app : java.lang.IllegalStateException: ContainerBase.addChild: start: org.apache.catalina.LifecycleException: org.apache.catalina.LifecycleException: org.glassfish.jersey.server.model.ModelValidationException: Validation of the application resource model has failed during application initialization.
[[FATAL] A resource model has ambiguous (sub-)resource method for HTTP method GET and input mime-types as defined by"@Consumes" and "@Produces" annotations at Java methods public abstract java.lang.String org.eclipse.microprofile.rest.client.tck.interfaces.cdi.scoped.StringClient.executeGet() and public abstract java.lang.String org.eclipse.microprofile.rest.client.tck.interfaces.cdi.scoped.StringClient.executeGet() at matching regular expression /. These two methods produces and consumes exactly the same mime-types and therefore their invocation as a resource methods will always fail.; source='org.glassfish.jersey.server.model.RuntimeResource@4811d82a', [FATAL] A resource model has ambiguous (sub-)resource method for HTTP method GET and input mime-types as defined by"@Consumes" and "@Produces" annotations at Java methods public abstract java.lang.String org.eclipse.microprofile.rest.client.tck.interfaces.cdi.scoped.StringClient.executeGet() and public abstract java.lang.String org.eclipse.microprofile.rest.client.tck.interfaces.cdi.scoped.StringClient.executeGet() at matching regular expression /. These two methods produces and consumes exactly the same mime-types and therefore their invocation as a resource methods will always fail.; source='org.glassfish.jersey.server.model.RuntimeResource@4811d82a', [FATAL] A resource model has ambiguous (sub-)resource method for HTTP method GET and input mime-types as defined by"@Consumes" and "@Produces" annotations at Java methods public abstract java.lang.String org.eclipse.microprofile.rest.client.tck.interfaces.cdi.scoped.StringClient.executeGet() and public abstract java.lang.String org.eclipse.microprofile.rest.client.tck.interfaces.cdi.scoped.StringClient.executeGet() at matching regular expression /. These two methods produces and consumes exactly the same mime-types and therefore their invocation as a resource methods will always fail.; source='org.glassfish.jersey.server.model.RuntimeResource@4811d82a']]]

[2025-08-27T00:25:33.277808+03:00] [GF 7.1.0-SNAPSHOT] [SEVERE] [] [org.glassfish.jersey.internal.Errors] [tid: _ThreadID=73 _ThreadName=admin-listener(1)] [levelValue: 1000] [[
  Following issues have been detected: 
WARNING: A resource model has ambiguous (sub-)resource method for HTTP method GET and input mime-types as defined by"@Consumes" and "@Produces" annotations at Java methods public abstract java.lang.String org.eclipse.microprofile.rest.client.tck.interfaces.cdi.scoped.StringClient.executeGet() and public abstract java.lang.String org.eclipse.microprofile.rest.client.tck.interfaces.cdi.scoped.StringClient.executeGet() at matching regular expression /. These two methods produces and consumes exactly the same mime-types and therefore their invocation as a resource methods will always fail.
WARNING: A resource model has ambiguous (sub-)resource method for HTTP method GET and input mime-types as defined by"@Consumes" and "@Produces" annotations at Java methods public abstract java.lang.String org.eclipse.microprofile.rest.client.tck.interfaces.cdi.scoped.StringClient.executeGet() and public abstract java.lang.String org.eclipse.microprofile.rest.client.tck.interfaces.cdi.scoped.StringClient.executeGet() at matching regular expression /. These two methods produces and consumes exactly the same mime-types and therefore their invocation as a resource methods will always fail.
WARNING: A resource model has ambiguous (sub-)resource method for HTTP method GET and input mime-types as defined by"@Consumes" and "@Produces" annotations at Java methods public abstract java.lang.String org.eclipse.microprofile.rest.client.tck.interfaces.cdi.scoped.StringClient.executeGet() and public abstract java.lang.String org.eclipse.microprofile.rest.client.tck.interfaces.cdi.scoped.StringClient.executeGet() at matching regular expression /. These two methods produces and consumes exactly the same mime-types and therefore their invocation as a resource methods will always fail.
]]
```

dmatej commented:
@arjantijms Any ideas?

avpinchuk commented:
This is a reproducer for `ClientClosedTest` deployment error: https://github.com/avpinchuk/rest-client

avpinchuk committed:
Rearrange manifest patch

Signed-off-by: Alexander Pinčuk <alexander.v.pinchuk@gmail.com>

avpinchuk commented:
@arjantijms, I tried with 2.1.1 everything works well.

With 2.1.0 and jersey 4.0.0-M3 the heap is still growing, but more slowly (OOME happen less often).

arjantijms commented:
That sounds good! Now we should maybe test 2.1.1 with jersey 4.0.0-m3 to see if that's again better / faster (less heap used)

dmatej committed:
Default TestNG version changed to 7.11.0

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
BeansXmlTransformer not required any more

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
The default heap size is ok now

- Higher was needed because of the arquillian gf container 2.1.0 mem leak, which
  was already fixed

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

avpinchuk commented:
> * Second is ClientClosedTest, I am trying to figure out why

@dmatej, this is a reproducer for ClientClosedTest deployment error: https://github.com/avpinchuk/rest-client

dmatej committed:
The beans.xml file is not needed any more

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Excluded challenged tests

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

arjantijms reviewed (approved):
Looks very good!
